### ISSUE TITLE
some question about codes of REINFORCE

### ISSUE THREAD

**Issue by zhiyiZeng** [boyu-ai/Hands-on-RL#7]
老师好，关于代码有点问题想请教。在REINFORCE类下的update函数，模型更新代码如下

```python

    def update(self, transition_dict):
        reward_list = transition_dict['rewards']
        state_list = transition_dict['states']
        action_list = transition_dict['actions']

        G = 0
        self.optimizer.zero_grad()
        for i in reversed(range(len(reward_list))):  # 从最后一步算起
            reward = reward_list[i]
            state = torch.tensor([state_list[i]], dtype=torch.float).to(self.device)
            action = torch.tensor([action_list[i]]).view(-1, 1).to(self.device)
            log_prob = torch.log(self.policy_net(state).gather(1, action))
            G = self.gamma * G + reward
            loss = -log_prob * G  # 每一步的损失函数
            loss.backward()  # 反向传播计算梯度
        self.optimizer.step()  # 梯度下降
```

我的问题：请问为何optimizer.step是放在循环外面而不是里面？以及loss.backward()为何放在里面而不是外面？

放在外面的话，相当于只更新最后一个循环的loss.backward()计算的梯度。那为何还要每个循环都执行loss.backward()，而不是把loss.backward()放在外面？

感谢解答~

**Comment by zerlinwang** [comment]
This is not an official answer!

- Putting `optimizer.step()` outside the loop can save calc cost (only once back-propagation)
- Putting `loss.backward()` outside is unreasonable (only last loss is used) and may result in params update slowly. It may be correct if you also change the code `loss = -log_prob * G` to `loss += -log_prob *G`.
- For more info, please read the pytorch doc.

**Comment by zhiyiZeng** [comment]
Thank you for your quicky reply. I see there's a point in changing `=` to `+=`. But the original code works suprisingly fine and I'm curious why.

I still don't understand why `optimizer.step()` is outside the loop. My understanding for `step()` is that it's used to actually update the parameters of the model. Therefore, in current codes, grad of loss will be calculated for multiple times but only update the parameters only in the last loop. So, why not calculate only once by moving outside the loop instead ? I don't think the original code is reasonable.

To make my self clear, I post the code from my understanding as follows:
```python
G = 0
for i in reversed(range(len(reward_list))):  # 从最后一步算起
    reward = reward_list[i]
    state = torch.tensor([state_list[i]], dtype=torch.float).to(self.device)
    action = torch.tensor([action_list[i]]).view(-1, 1).to(self.device)
    log_prob = torch.log(self.policy_net(state).gather(1, action))
    G = self.gamma * G + reward
    loss = -log_prob * G  # 每一步的损失函数
    self.optimizer.zero_grad()
    loss.backward()  # 反向传播计算梯度
    self.optimizer.step()  # 梯度下降
```

**Comment by zhiyiZeng** [comment]
> * Putting `loss.backward()` outside is unreasonable (only last loss is used) and may result in params update slowly. It may be correct if you also change the code `loss = -log_prob * G` to `loss += -log_prob *G`.

I tried the way you suggest(code as follows). The rewards are high but the training time is far much longer(over 10 mins for the fist 6 epochs). So, I don't think the suggestion is what the authors originally thought of.
```python
    def update(self, transition_dict):
        reward_list = transition_dict['rewards']
        state_list = transition_dict['states']
        action_list = transition_dict['actions']

        # NOTE optimizer.step()放在循环外面，意思是，只有第一步的时候才更新。似乎也是合理的，让网络更加robust。
        G = 0
        loss = 0
        self.optimizer.zero_grad()
        for i in reversed(range(len(reward_list))):  # 从最后一步算起
            reward = reward_list[i]
            state = torch.tensor([state_list[i]], dtype=torch.float).to(self.device)
            action = torch.tensor([action_list[i]]).view(-1, 1).to(self.device)
            log_prob = torch.log(self.policy_net(state).gather(1, action))
            G = self.gamma * G + reward
            loss += -log_prob * G  # 每一步的损失函数
            loss.backward(retain_graph=True)  # 反向传播计算梯度
        self.optimizer.step()  # 梯度下降
```

**Comment by zhiyiZeng** [comment]
I see, thank you!
